GraphWebApi/Telemetry/TelemetryMiddleware.cs

```diff
@@ -0,0 +1,102 @@
+﻿// ------------------------------------------------------------------------------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------------------------------------------------------------------------------
+
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.AspNetCore.Http;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace GraphWebApi.Telemetry
+{
+    /// <summary>
+    /// This class implements a telemetry middleware for correlating telemetry operations
+    /// </summary>
+    public class TelemetryMiddleware
+    {
+        /* you may create a new TelemetryConfiguration instance, reuse one you already have 
+         * or fetch the instance created by Application Insights SDK.
+        */
+        private readonly RequestDelegate _next;
+
+        public TelemetryMiddleware(RequestDelegate next)
+        {
+            _next = next
+                ?? throw new ArgumentNullException(nameof(next), $"{ next }: { nameof(next) }");
+        }
+
+        public async Task Invoke(HttpContext context, TelemetryClient client)
+        {
+            if(context != null)
+            {
+                // Let's create and start RequestTelemetry.
+                var requestTelemetry = new RequestTelemetry
+                {
+                    // "Request {method} {url} => {statusCode}" format
+                    Name = $"{context.Request.Method + " " + context.Request.Path.Value}"
+                };
+                requestTelemetry.Properties.Add("HttpMethod", context.Request.Method);
+
+                var requestHeaders = context.Request.Headers;
+
+                // If there is a Request-Id received from the upstream service, set the telemetry context accordingly.
+                if (requestHeaders.ContainsKey("Request-Id"))
+                {
+                    var requestId = requestHeaders["Request-Id"];
+
+                    // Get the operation ID from the Request-Id (if you follow the HTTP Protocol for Correlation).
+                    requestTelemetry.Context.Operation.Id = GetOperationId(requestId);
+                    requestTelemetry.Context.Operation.ParentId = requestId;
+                }
+
+                // StartOperation is a helper method that allows correlation of 
+                // current operations with nested operations/telemetry
+                // and initializes start time and duration on telemetry items.
+                using (var operation = client.StartOperation(requestTelemetry))
```

Comment:
You may simplify the using statement:

```suggestion
                using var operation = client.StartOperation(requestTelemetry);
```
Then remove the redundant open and close braces that follow.